examples/stabilizer_code_test.py

```diff
@@ -3,37 +3,60 @@
 import examples.stabilizer_code as sc
 
 
-def encode_corrupt_correct(input_val, error_gate, error_loc):
-    code = sc.StabilitizerCode(group_generators=['XZZXI', 'IXZZX', 'XIXZZ', 'ZXIXZ'])
-
+def encode_corrupt_correct(code, input_val, error_gate, error_loc):
     circuit = cirq.Circuit()
-    qubits = [cirq.NamedQubit(name) for name in ['0', '1', '2', '3', 'c']]
-    ancillas = [cirq.NamedQubit(name) for name in ['d0', 'd1', 'd2', 'd3']]
+    qubits = [cirq.NamedQubit(str(i)) for i in range(code.n - code.k)] + [cirq.NamedQubit('c')]
+    ancillas = [cirq.NamedQubit(f"d{i}") for i in range(code.n - code.k)]
+
+    circuit += code.encode(qubits)
 
-    code.encode(circuit, qubits)
     if error_gate and error_loc:
         circuit.append(error_gate(qubits[error_loc]))
 
-    code.correct(circuit, qubits, ancillas)
+    circuit += code.correct(qubits, ancillas)
 
     results = cirq.Simulator().simulate(
-        circuit, qubit_order=(qubits + ancillas), initial_state=(input_val * 2 ** 4)
+        circuit, qubit_order=(qubits + ancillas), initial_state=(input_val * 2 ** len(ancillas))
     )
 
-    qubit_map = {qubit: i for i, qubit in enumerate(qubits + ancillas)}
-    pauli_string = cirq.PauliString(dict(zip(qubits, code.Z[0])))
-    trace = pauli_string.expectation_from_state_vector(results.state_vector(), qubit_map)
+    decoded = code.decode(qubits, ancillas, results.state_vector())
 
-    return round((1 - trace.real) / 2)
+    return decoded[0]
 
 
 def test_no_error():
+    # Table 3.2.
+    five_qubit_code = sc.StabilizerCode(
+        group_generators=['XZZXI', 'IXZZX', 'XIXZZ', 'ZXIXZ'], allowed_errors=['X', 'Z']
+    )
+
     for input_val in [0, 1]:
-        assert encode_corrupt_correct(input_val, None, None) == input_val
+        assert encode_corrupt_correct(five_qubit_code, input_val, None, None) == input_val
 
 
 def test_errors():
+    # Table 3.2.
+    five_qubit_code = sc.StabilizerCode(
+        group_generators=['XZZXI', 'IXZZX', 'XIXZZ', 'ZXIXZ'], allowed_errors=['X', 'Z']
+    )
+
     for input_val in [0, 1]:
         for error_gate in [cirq.X, cirq.Z]:
-            for error_loc in range(5):
-                assert encode_corrupt_correct(input_val, error_gate, error_loc) == input_val
+            for error_loc in range(five_qubit_code.n):
+                assert (
+                    encode_corrupt_correct(five_qubit_code, input_val, error_gate, error_loc)
+                    == input_val
+                )
+
+
+def test_imperfect_code():
+    # Also known as the bit-flip code.
+    seven_qubit_code = sc.StabilizerCode(group_generators=['ZZI', 'ZIZ'], allowed_errors=['X'])
+
+    for error_gate in [cirq.X]:
```

Comment:
Would be interesting to also test here that `cirq.Z` errors are _not_ corrected by this code.

Reply:
So I am having trouble with this, because the current code does seem to correct these errors... I debugged the code:
https://github.com/tonybruguier/sandbox/blob/master/debug_ecc.txt

I printed the state vector right after calling `simulate()`, and I think the code itself behaves correctly. The first 3 / 5 qubits are the code words and the last 2/4 ancillas are the syndromes. 

We I do see that the Z errors result in an incorrect state vector only for the bit flip code, but the Pauli string measurements do not capture this. I think this is because there is a global phase. How do you think I should test?

Leaving unresolved, because I think there's more work to do.